compiler/rustc_target/src/abi/mod.rs

```diff
@@ -803,25 +822,16 @@ pub struct Scalar {
 impl Scalar {
     #[inline]
     pub fn is_bool(&self) -> bool {
-        matches!(self.value, Int(I8, false))
-            && matches!(self.valid_range, WrappingRange { start: 0, end: 1 })
+        matches!(
+            self,
+            Scalar { value: Int(I8, false), valid_range: WrappingRange { start: 0, end: 1 } }
+        )
     }
 
-    /// Returns the valid range as a `x..y` range.
-    ///
-    /// If `x` and `y` are equal, the range is full, not empty.
-    pub fn valid_range_exclusive<C: HasDataLayout>(&self, cx: &C) -> Range<u128> {
-        // For a (max) value of -1, max will be `-1 as usize`, which overflows.
-        // However, that is fine here (it would still represent the full range),
-        // i.e., if the range is everything.
-        let bits = self.value.size(cx).bits();
-        assert!(bits <= 128);
-        let mask = !0u128 >> (128 - bits);
-        let start = self.valid_range.start;
-        let end = self.valid_range.end;
-        assert_eq!(start, start & mask);
-        assert_eq!(end, end & mask);
-        start..(end.wrapping_add(1) & mask)
+    /// Returns `true` if all possible numbers are valid, i.e `valid_range` covers the whole layout
+    #[inline]
+    pub fn is_always_valid_for<C: HasDataLayout>(&self, cx: &C) -> bool {
+        self.valid_range.is_full_for(self.value.size(cx))
     }
```

Comment:
I like this setup! Only one tiny nit, I don't think `is_always_valid_for` needs the `_for` at the end. The argument doesn't dictate validity necessarily, it's just a required context. So `is_always_valid` seems enough.

As a complete aside (and a bad idea to tackle in this PR, heh), I wonder how many context arguments on methods we could remove if we got rid of the sole `Primitive::Pointer` and actually made it track its own size.

Reply:
I actually thought about how much happier my life would be if `Primitive::Pointer` was gone.
It seems very entrenched though, is it even possible to remove?

Reply:
We can't remove it fully (there are semantic differences), only make it so `TargetDataLayout` isn't needed to compute its size.

Reply:
I'm confused, I thought the size of the pointer would change depending on context?

Reply:
Working on something like this.
Is it reasonable from a performance perspective to store `Size` in `Primitive`?
`./x.py test ui` passes